sql/core/src/main/scala/org/apache/spark/sql/execution/CacheManager.scala

```diff
@@ -163,12 +161,15 @@ class CacheManager extends Logging {
     val relation = cachedData.cachedRepresentation
     val (rowCount, newColStats) =
       CommandUtils.computeColumnStats(sparkSession, relation, column)
-    val oldStats = relation.statsOfPlanToCache
-    val newStats = oldStats.copy(
-      rowCount = Some(rowCount),
-      attributeStats = AttributeMap((oldStats.attributeStats ++ newColStats).toSeq)
-    )
-    relation.statsOfPlanToCache = newStats
+
+    relation.synchronized {
```

Comment:
I would move this logic into the class `InMemoryRelation`. Like:

```scala
def updateStats(rowCount: Long, newColStats: Map[Attribute, ColumnStat]): Unit = synchronized {
    val newStats = statsOfPlanToCache.copy(
      rowCount = Some(rowCount),
      attributeStats = AttributeMap((statsOfPlanToCache.attributeStats ++ newColStats).toSeq)
    )
    statsOfPlanToCache = newStats
  }
```

This way both the write and the `synchronized` is in the member function of the `InMemoryRelation` (I think it is bit more readable). What is your opinion?

Reply:
Looks reasonable, I'll update soon.